Make empty_cpu sanity test CPU only in DEBUG mode

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51360 Test if allocator is set only in DEBUG mode.
* **#51358 Make empty_cpu sanity test CPU only in DEBUG mode**

BackendSelect is expected to enforce this invariant.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26149502](https://our.internmc.facebook.com/intern/diff/D26149502)

ezyang committed:
Make empty_cpu sanity test CPU only in DEBUG mode

BackendSelect is expected to enforce this invariant.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

swolchok reviewed (approved):
Seems fine. Why accept the device as an argument at all then?

ezyang commented:
@swolchok Hmm, while this is a very valid point, I'm not sure how to do it in a reasonable way. Device shows up in the backend implementations because backend implementations have the same type signature as the interfaces they implement, and in the interface it isn't known what the device is.

ezyang committed:
Update on "Make empty_cpu sanity test CPU only in DEBUG mode"


BackendSelect is expected to enforce this invariant.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26149502](https://our.internmc.facebook.com/intern/diff/D26149502)

[ghstack-poisoned]
